tools/sourcerepo-sync/source_sync.py

```diff
@@ -46,93 +53,107 @@ def main() -> None:
   if args.verbose:
     print(f'Loaded {len(local_sourcerepos)} local source repositories')
 
+  # Validate the yaml file
   # Check sourcerepo for duplicates:
-  sourcerepo_names = []
-  for repo in local_sourcerepos:
-    if repo['name'] == '' or repo['name'] is None or repo['name'] == 'null':
-      raise ValueError(f'Empty sourcerepo name in {file}')
-    if repo['name'] in sourcerepo_names:
-      raise ImportError(f'Duplicate sourcerepo name {repo["name"]} in {file}')
-    sourcerepo_names.append(repo['name'])
-
-  client = datastore.Client(project=args.project)
-  query = client.query(kind=args.kind)
-
-  ds_repos = list(query.fetch())
-  if args.verbose:
-    print(f'Retrieved {len(ds_repos)} source repositories from datastore')
-
-  for repo in local_sourcerepos:
-    repo_found = False
-    for ds_repo in ds_repos:
-      # if it exists, check if it needs to be updated
-      if repo['name'] == ds_repo['name']:
-        repo_found = True
-        if args.verbose:
-          print(f'Found source repository {repo["name"]}')
-        ds_repos.pop(ds_repos.index(ds_repo))
-        update_sourcerepo(repo, ds_repo, args, client, args.kind)
-
-    if repo_found:
-      continue
-    create_sourcerepo(repo, args, client, args.kind)
+  def validate_repository(repository, local=True):
+    """Check the attributes of the source repo."""
+    sourcerepo_names = []
+    for repo in repository:
+      if repo['name'] == '' or repo['name'] is None or repo['name'] == 'null':
+        raise ValueError(f'Empty sourcerepo name in {file}')
+      if repo['name'] in sourcerepo_names:
+        raise ImportError(f'Duplicate sourcerepo name {repo["name"]} in {file}')
+      sourcerepo_names.append(repo['name'])
+      # Make sure the link ends with a /
+      if 'link' in repo and repo['link'][-1] != '/':
+        raise ValueError(f'Link in {repo["name"]} missing ending /')
+      # Check for a dynamic field being mistakenly pushed
+      if local:
+        dynamic_fields = {
+            'last_update_date', 'ignore_last_import_time', 'last_synced_hash'
+        }
+        for field in dynamic_fields:
+          if field in repo:
+            raise ValueError(f'Dynamic field {field} found in {repo["name"]}')
+    if args.verbose:
+      print(f'Validated {len(repository)} source repositories')
 
-  local_sourcerepos_names = {repo['name'] for repo in local_sourcerepos}
-  # If the source repo is not in the local yaml, delete it
-  for ds_repo in ds_repos:
-    if ds_repo['name'] not in local_sourcerepos_names:
+  def create_sourcerepo(repo):
+    """Create a new source repo."""
+    if args.file.startswith('../../'):
```

Comment:
I think it's better to check for if the working directory is `sourcerepo-sync` instead.

Reply:
Done.